lib/spack/spack/bootstrap.py

```diff
@@ -28,6 +31,208 @@
 import spack.util.path
 from spack.util.environment import EnvironmentModifications
 
+#: Map a bootstrapper type to the corresponding class
+_bootstrap_methods = {}
+
+
+def _bootstrapper(type):
+    """Decorator to register classes implementing bootstrapping
+    methods.
+
+    Args:
+        type (str): string identifying the class
+    """
+    def _register(cls):
+        _bootstrap_methods[type] = cls
+        return cls
+    return _register
+
+
+@_bootstrapper(type='store')
```

Comment:
I think these classes should be slightly refactored -- The latter two should inherit from `_StoreBootstrapper` -- they all bootstrap into the bootstrap store

Then they can inherit the `try_import` method and use that as they need to. The `_BuildcacheBootstrapper` and `_SourceBootstrapper` should then have an additional method `provide_module` that contains the logic currently under `try_import` in those modules, and `_StoreBootstrapper.provide_module` should be a simple call to `_StoreBootstrapper.try_import`, since that's the only way it can provide it.

Then `ensure_module_importable_or_raise` can call `b.provide_module` for each module, and we won't have to have non-hierarchical class references in _SourceBootstrapper and _BuildcacheBootstrapper/

Reply:
> Then ensure_module_importable_or_raise can call b.provide_module for each module, and we won't have to have non-hierarchical class references in _SourceBootstrapper and _BuildcacheBootstrapper/

I understand the concern here, but tried a different solution in aeaf871. I didn't go with a class hierarchy since it seemed an overkill to me (all the classes I can think of right now would be "final" and without mixins). ` _SourceBootstrapper` is basically an adapter to avoid adding more complexity to the controller function here:

https://github.com/spack/spack/blob/2e647d40f55affdc7eaf69cb93cbc869c724fca7/lib/spack/spack/bootstrap.py#L292-L308

but I was lazy enough to avoid extracting the code from it into a standalone function. Not there's a `_try_import_from_store` function called by the three classes. Let me know if this is enough to solve the issue.

Reply:
I think we should be able to remove the `_StoreBootstrapper` class entirely by calling `_try_import_from_store` at the top of the `try_import` methods. That leads to checking the store an extra time per-bootstrapper (which we could avoid with careful addition of a parameter that says we've already checked and know it's missing), but it would allow us to get rid of the ugly bits of code around patching the `_StoreBootstrapper` and its trust information into the configs that we read in. How do you feel about that idea?

Reply:
Done in f8b13e4